clang-format: make formatting deterministic for different formatter versions

l0rinc opened the pull request:
Updates `.clang-format` file to reflect [latest supported Clang-Format standards](https://releases.llvm.org/16.0.0/tools/clang/docs/ClangFormatStyleOptions.html) while preserving most of the existing formatting behavior.

Note that [`AfterStruct` brace placement](https://github.com/bitcoin/bitcoin/pull/32414#discussion_r2072678126) was originally aligned here with `AfterClass`, but was reverted by reviewer demand.

maflcko reviewed (commented):
last commit seems fine, but the others i am not sure. at a minimum you'll have to drop the scripted diffs from executing, because they can't be  reproduced anyway with different clang-format versions

l0rinc commented on src/.clang-format:
```diff
@@ -1,26 +1,40 @@
 AccessModifierOffset: -4
-AlignAfterOpenBracket: true
+AlignAfterOpenBracket: Align
```
Many of these keys aren't even listed in the documentation and the IDE shows that many are invalid since it doesn't match the allowed schema.

l0rinc commented on src/.clang-format:
```diff
@@ -1,49 +1,26 @@
-BreakBeforeConceptDeclarations: Always
-RequiresExpressionIndentation: OuterScope
-Standard: c++20
-UseTab:          Never
```
I've added all of them, this was the simplest change with fewest commits and explanations

l0rinc commented on src/.clang-format:
```diff
@@ -1,49 +1,222 @@
-AlignAfterOpenBracket: true
-AlignEscapedNewlinesLeft: true
-AlignTrailingComments: true
+AlignAfterOpenBracket: Align
```
> How to reproduce the commit on non-macos

Please see the commit messages - should I move the instructions to the PR description as well?

> produces a different result

I think you have dumped the defaults that don't load our config first, which are indeed different from what we have (hence our own config).
```bash
$ clang-format-16 --version
Ubuntu clang-format version 16.0.6 (23ubuntu4)
$ clang-format-16 -dump-config | grep AfterClass
  AfterClass:      false
$ clang-format-16 -dump-config -style=file:src/.clang-format | grep AfterClass
  AfterClass:      true
```

and

```bash
clang-format-16 -dump-config -style=file:src/.clang-format | egrep 'AccessModifierOffset|AlignEscapedNewlines|AllowShortCaseLabelsOnASingleLine|AllowShortIfStatementsOnASingleLine|AlwaysBreakTemplateDeclarations|AfterClass|AfterFunction|AfterStruct|BreakBeforeBraces|BreakBeforeTernaryOperators|ColumnLimit|IndentWidth|KeepEmptyLinesAtTheStartOfBlocks|MaxEmptyLinesToKeep|PointerAlignment|Standard'
AccessModifierOffset: -4
AlignEscapedNewlines: Left
AllowShortCaseLabelsOnASingleLine: true
AllowShortIfStatementsOnASingleLine: WithoutElse
AlwaysBreakTemplateDeclarations: Yes
  AfterClass:      true
  AfterFunction:   true
  AfterStruct:     false
BreakBeforeBraces: Custom
BreakBeforeTernaryOperators: false
ColumnLimit:     0
ConstructorInitializerIndentWidth: 4
ContinuationIndentWidth: 4
DerivePointerAlignment: false
IndentWidth:     4
KeepEmptyLinesAtTheStartOfBlocks: false
MaxEmptyLinesToKeep: 2
ObjCBlockIndentWidth: 2
PointerAlignment: Left
PPIndentWidth:   -1
  AfterFunctionDefinitionName: false
  AfterFunctionDeclarationName: false
Standard:        c++20
```

maflcko reviewed (approved):
I haven't checked if this fixes the issues with your IDE, but I guess it makes sense either way to get the clang-format config up to date. clang-16 is the minimum required for about a year now, so every developer should have access to clang-format-16 in some way or another.

I've recreated the commits on Ubuntu and reviewed the diff:

review ACK 94364a7d5403d9d480ebc065f8709c6dd21e543f 🥙

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 94364a7d5403d9d480ebc065f8709c6dd21e543f 🥙
xsnBRKa1+85+2JOjTRQ1kZgSlBHhVZxj7LlQpMQNkd5vPsoJ/6a84fiaWp9pAnna9VGz0/DA4J3jNoByA1TuCg==
```

</details>

hodlinator commented on doc/developer-notes.md:
```diff
@@ -67,7 +67,7 @@ Coding Style (C++)
 [clang-format-diff script](/contrib/devtools/README.md#clang-format-diffpy)
 tool to clean up patches automatically before submission.
-  - Braces on new lines for classes, functions, methods.
+  - Braces on new lines for classes, structs, functions, methods.
```
Why change the C++ Coding Style? I got used to the current one, which encourages `struct`s for POD as a separate thing.

l0rinc replied:
We can define it that way as well (though I don't agree with the distinction vs classes), but not specifying it directly results in different versions formatting structs differently, so we should define it explicitly

l0rinc replied:
I would be fine with both as long as it's explicit. I still find the rationale that `struct` and `class` should be formatted similarly, it's harder for me to explain why `struct` is more like a for loop than a `class`...

> 85'849 lines 

I have no idea what these numbers mean, I checked same-line opening braces with
```bash
% grep -E '^\s*\bstruct\b\s+[^;{]*\{' --include="*.h" --include="*.cpp" -r src/ | wc -l
     393
```

and quickly vibe coded a script that does newlines:

<details>
<summary>vibe-Details</summary>

```bash
#!/bin/bash

echo "Analyzing struct brace placement in Bitcoin Core..."

# Count same-line braces
same_line=$(grep -E '^\s*struct\s+[^;{]*\{' --include="*.h" --include="*.cpp" -r src/ | wc -l)

# Count next-line braces using awk
next_line=$(find src/ \( -name "*.cpp" -o -name "*.h" \) -exec awk '
/^[[:space:]]*struct[[:space:]]+[^;{]*$/ {
    struct_line = NR
    getline
    if (/^[[:space:]]*\{/) {
        count++
    }
}
END { print count+0 }
' {} \; | awk '{sum+=$1} END {print sum}')

echo "Structs with same-line brace: $same_line"
echo "Structs with next-line brace: $next_line"
echo "Total: $((same_line + next_line))"
``

</details>

which vibe-revealed:
```
Analyzing struct brace placement in Bitcoin Core...
Structs with same-line brace:      393
Structs with next-line brace: 223
Total: 616
```

So the counts don't really matter in my opinion, we have a few hundred of each, there's no clear winner based on usage, let's come up with better arguments for either.

l0rinc replied:
> Admittedly they are imprecise

I'd say ~85000 vs ~400 is a bit more than imprecise :)

> almost double on same line

That's not how we usually decide outcomes, rather by better explanations - especially since many of those aren't counting declarations, e.g
* https://github.com/bitcoin/bitcoin/blob/f7d2db28e90297664390d527881ebbbf2c6ce6f0/src/bitcoin-cli.cpp#L241
* https://github.com/bitcoin/bitcoin/blob/face8123fdc10549676c6679ee3225c178a7f30c/src/httpserver.cpp#L565
* https://github.com/bitcoin/bitcoin/blob/face8123fdc10549676c6679ee3225c178a7f30c/src/httpserver.cpp#L579
* https://github.com/bitcoin/bitcoin/blob/f58de8749e94c4f6d5ab4a13d098ccfc291336c7/src/net.cpp#L3162

I'm fine with both, but have better explanations for the one I'm proposing. Ultimately I don't think it matters much, as long as it's explicit since current behavior results in inconsistencies.

> Maybe seeing struct as a small/passive/record thing and seeing class as more of an alive/process

That sounds more philosophical to me, I don't think that's how we're using them (or how formatting is usually decided)? Most definitions I see state: it's just a class with public members, e.g. https://en.wikipedia.org/wiki/Struct_(C_programming_language)#C++
> a class is the same as a struct but with different default visibility

Not sure why the formatting should be closer to an `if` or a `switch` or a loop, I don't see how it's closer to those that to a class.

Clang-format itself usually bundles the class and struct formatting together:
* https://github.com/llvm/llvm-project/blob/d397751b56e30d88e3f00464c5a5fdfb2771978d/clang/lib/Format/TokenAnnotator.cpp#L5943-L5947
* https://github.com/llvm/llvm-project/blob/d397751b56e30d88e3f00464c5a5fdfb2771978d/clang/lib/Format/UnwrappedLineFormatter.cpp#L492-L495

hodlinator replied:
> especially since many of those aren't counting declarations, e.g ...

That is precisely one of the reasons why running clang-format with the two different values on the entire src/ and comparing the relative diff size is helpful. Lines such as the ones you quoted will not contribute to the relative difference in the size of the change.

---

> That's not how we usually decide outcomes, rather by better explanations - 

I try to present explanations, but then you respond with:

> That sounds more philosophical to me

But I've been brewing on a variation of the argument. Say you have a C coding style with `struct Foo {` and `function\n{`. Then you start introducing C++. Suddenly you have these struct-like classes but they can contain functions. C++ came from Bjarne's "C with classes" https://en.wikipedia.org/wiki/Cfront. `class` could contain functions and so allowing `class Foo\n{` would seem more okay. `struct`s could also (eventually?) contain functions since it was sometimes useful. I know this argument is somewhat blunt, but I think our current style is influenced by the historical progression of C++ coming out of C.

---

Elaborating on my libsecp256k1 argument since you did not respond to it:
sipa both added the section to the coding style and wrote that library to replace OpenSSL-usage. They both happened within a year of each other. Keeping styles consistent between the two closely related project is nice for developers working on both (silent payments, musig2 and batch signature verification comes to mind as projects which span both projects). So even if our `struct` style is understated, I would wager that your change to developer-notes.md constitutes a *change of style* rather than a clarification or specifying something that was totally arbitrary.

l0rinc replied:
I didn't mean my response to be offensive before, and while still not sure I fully follow the arguments, I agree that stabilizing the formatting and unifying class/struct braces are separate concerns, so I have removed the unification, only kept the change making formatters explicit.

hodlinator reviewed (commented):
ACK 7c9b3e1eae8f206753457149f1b1c837f6627d6d (first commit)

Ran `clang-format -dump-config -style=file:src/.clang-format > foo` on that commit and diffed *foo* against *src/.clang-format*. Only differences where added/elaborated settings from me running Clang-format 19. Makes sense to keep the format at the minimum version we support.

l0rinc committed:
clang-format: regenerate configs

Regenerated `.clang-format` from current configs to replace deprecated keys with up-to-date equivalents.
Also added all current formatter default values to guard against version differences.

The configs were updated with the following command (using v16 for maximal compatibility):
   $(brew --prefix llvm@16)/bin/clang-format -dump-config -style=file:src/.clang-format

The new config was tested with:
   $(brew --prefix llvm@16)/bin/clang-format -i src/deploymentinfo.h

Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

hodlinator reviewed (approved):
re-ACK 13f36c020f0329b5e975282b45292fdf2a495e31

`struct` brace style was made to explicitly differ from `class` in developer-notes.md after some back & forth (https://github.com/bitcoin/bitcoin/pull/32813#discussion_r2303151367).
